### ISSUE TITLE
The addon ip-masq-agent could not be disabled for Kubernetes 1.11.5

### ISSUE THREAD

**Issue by jayanthpurushothaman** [Azure/aks-engine#160]
**Is this a request for help?**:
NO

**Is this an ISSUE or FEATURE REQUEST?** (choose one):
ISSUE

**What version of acs-engine?**:
0.26.2

**Orchestrator and version (e.g. Kubernetes, DC/OS, Swarm)**
Kubernetes 1.11.5

**What happened**:
While running `acs-engine generate` command with the following stanza in the `kubernetesConfig` section of the cluster definition JSON, we get the error shown below, only if `ip-masq-agent` is disabled. Otherwise it doesn't throw any error:

**Addons stanza**
```
        "addons": [
          {
            "name": "blobfuse-flexvolume",
            "enabled" : false
          },
          {
            "name": "smb-flexvolume",
            "enabled" : false
          },
          {
            "name": "keyvault-flexvolume",
            "enabled" : false
          },
          {
            "name": "nvidia-device-plugin",
            "enabled" : false
          },
          {
            "name": "azure-cni-networkmonitor",
            "enabled" : false
          },
          {
           "name": "ip-masq-agent",
           "enabled" : false
          }
        ]
```

**Error**
```
INFO[0000] Generating assets into _output/test...
FATA[0011] error generating template template.json: runtime error: index out of range - goroutine 1 [running]:
runtime/debug.Stack(0xc000023580, 0x1f81240, 0x2ee6fd0)
	/usr/local/go/src/runtime/debug/stack.go:24 +0xa7
github.com/Azure/acs-engine/pkg/acsengine.(*TemplateGenerator).GenerateTemplate.func2(0xc000025b28, 0xc000025b08, 0xc000025b18)
	/go/src/github.com/Azure/acs-engine/pkg/acsengine/template_generator.go:78 +0x6e
panic(0x1f81240, 0x2ee6fd0)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
text/template.errRecover(0xc0000258e8)
	/usr/local/go/src/text/template/exec.go:160 +0x1d0
panic(0x1f81240, 0x2ee6fd0)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
text/template.errRecover(0xc000024388)
	/usr/local/go/src/text/template/exec.go:160 +0x1d0
panic(0x1f81240, 0x2ee6fd0)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/Azure/acs-engine/pkg/acsengine.getAddonFuncMap.func1(0xc00051a9da, 0xd, 0x0, 0x0)
	/go/src/github.com/Azure/acs-engine/pkg/acsengine/engine.go:700 +0xb6
reflect.Value.call(0x1f34760, 0xc0000ba0f0, 0x13, 0x21011e2, 0x4, 0xc000314120, 0x1, 0x1, 0x20fb200, 0x3391401, ...)
	/usr/local/go/src/reflect/value.go:447 +0x449
reflect.Value.Call(0x1f34760, 0xc0000ba0f0, 0x13, 0xc000314120, 0x1, 0x1, 0x22c1ae0, 0xc000102880, 0x1f0ece0)
	/usr/local/go/src/reflect/value.go:308 +0xa4
text/template.(*state).evalCall(0xc000024308, 0x2049e20, 0xc0000ba2d0, 0x99, 0x1f34760, 0xc0000ba0f0, 0x13, 0x22c1860, 0xc00035af00, 0xc00051a9ca, ...)
	/usr/local/go/src/text/template/exec.go:695 +0x6bd
text/template.(*state).evalFunction(0xc000024308, 0x2049e20, 0xc0000ba2d0, 0x99, 0xc00035af30, 0x22c1860, 0xc00035af00, 0xc000314020, 0x2, 0x2, ...)
	/usr/local/go/src/text/template/exec.go:563 +0x172
text/template.(*state).evalCommand(0xc000024308, 0x2049e20, 0xc0000ba2d0, 0x99, 0xc00035af00, 0x1f81540, 0x2f1a218, 0x99, 0x2c8, 0xc000668900, ...)
	/usr/local/go/src/text/template/exec.go:460 +0x5e9
text/template.(*state).evalPipeline(0xc000024308, 0x2049e20, 0xc0000ba2d0, 0x99, 0xc0004f5aa0, 0x0, 0x0, 0xc000024208)
	/usr/local/go/src/text/template/exec.go:429 +0x126
text/template.(*state).walk(0xc000024308, 0x2049e20, 0xc0000ba2d0, 0x99, 0x22c17a0, 0xc00035af60)
	/usr/local/go/src/text/template/exec.go:254 +0x49b
text/template.(*state).walk(0xc000024308, 0x2049e20, 0xc0000ba2d0, 0x99, 0x22c19a0, 0xc00035aea0)
	/usr/local/go/src/text/template/exec.go:262 +0x142
text/template.(*Template).execute(0xc000102480, 0x22a5000, 0xc00029a310, 0x2049e20, 0xc0000ba2d0, 0x0, 0x0)
	/usr/local/go/src/text/template/exec.go:217 +0x215
text/template.(*Template).Execute(0xc000102480, 0x22a5000, 0xc00029a310, 0x2049e20, 0xc0000ba2d0, 0x0, 0x0)
	/usr/local/go/src/text/template/exec.go:200 +0x53
github.com/Azure/acs-engine/pkg/acsengine.getContainerAddonsString(0xc0001e8840, 0x210f4e8, 0x13, 0x0, 0x0)
	/go/src/github.com/Azure/acs-engine/pkg/acsengine/engine.go:759 +0x611
github.com/Azure/acs-engine/pkg/acsengine.(*TemplateGenerator).getMasterCustomData(0xc0002ae2b8, 0xc0001fd5c0, 0x212371c, 0x22, 0xc0001e8840, 0x16, 0xc0001ff660)
	/go/src/github.com/Azure/acs-engine/pkg/acsengine/template_generator.go:196 +0x3bb
github.com/Azure/acs-engine/pkg/acsengine.(*TemplateGenerator).getTemplateFuncMap.func59(0xc0001e8840, 0x0, 0x0)
	/go/src/github.com/Azure/acs-engine/pkg/acsengine/template_generator.go:564 +0x52
reflect.Value.call(0x1f2b220, 0xc00000b8e0, 0x13, 0x21011e2, 0x4, 0xc0001ff600, 0x1, 0x1, 0x20fb200, 0x1eeee01, ...)
	/usr/local/go/src/reflect/value.go:447 +0x449
reflect.Value.Call(0x1f2b220, 0xc00000b8e0, 0x13, 0xc0001ff600, 0x1, 0x1, 0x22c18a0, 0xc00017aec0, 0x20ccb60)
	/usr/local/go/src/reflect/value.go:308 +0xa4
text/template.(*state).evalCall(0xc000025300, 0x20ccb60, 0xc0001e8840, 0x16, 0x1f2b220, 0xc00000b8e0, 0x13, 0x22c1860, 0xc0001eab10, 0xc0004aaf28, ...)
	/usr/local/go/src/text/template/exec.go:695 +0x6bd
text/template.(*state).evalFunction(0xc000025300, 0x20ccb60, 0xc0001e8840, 0x16, 0xc0001eab40, 0x22c1860, 0xc0001eab10, 0xc00017aee0, 0x2, 0x2, ...)
	/usr/local/go/src/text/template/exec.go:563 +0x172
text/template.(*state).evalCommand(0xc000025300, 0x20ccb60, 0xc0001e8840, 0x16, 0xc0001eab10, 0x1f81540, 0x2f1a218, 0x99, 0xc000024fc0, 0x1315c69, ...)
	/usr/local/go/src/text/template/exec.go:460 +0x5e9
text/template.(*state).evalPipeline(0xc000025300, 0x20ccb60, 0xc0001e8840, 0x16, 0xc000681140, 0x0, 0x0, 0x20ccb60)
	/usr/local/go/src/text/template/exec.go:429 +0x126
text/template.(*state).walk(0xc000025300, 0x20ccb60, 0xc0001e8840, 0x16, 0x22c17a0, 0xc0001eab70)
	/usr/local/go/src/text/template/exec.go:254 +0x49b
text/template.(*state).walk(0xc000025300, 0x20ccb60, 0xc0001e8840, 0x16, 0x22c19a0, 0xc0001eaab0)
	/usr/local/go/src/text/template/exec.go:262 +0x142
text/template.(*state).walkIfOrWith(0xc000025300, 0xa, 0x20ccb60, 0xc0001e8840, 0x16, 0xc0006810e0, 0xc0001eaab0, 0x0)
	/usr/local/go/src/text/template/exec.go:292 +0x1ca
text/template.(*state).walk(0xc000025300, 0x20ccb60, 0xc0001e8840, 0x16, 0x22c1960, 0xc00023fa80)
	/usr/local/go/src/text/template/exec.go:259 +0x327
text/template.(*state).walk(0xc000025300, 0x20ccb60, 0xc0001e8840, 0x16, 0x22c19a0, 0xc00067bb60)
	/usr/local/go/src/text/template/exec.go:262 +0x142
text/template.(*state).walkTemplate(0xc000025868, 0x20ccb60, 0xc0001e8840, 0x16, 0xc0002bf640)
	/usr/local/go/src/text/template/exec.go:411 +0x23e
text/template.(*state).walk(0xc000025868, 0x20ccb60, 0xc0001e8840, 0x16, 0x22c1b20, 0xc0002bf640)
	/usr/local/go/src/text/template/exec.go:267 +0x205
text/template.(*state).walk(0xc000025868, 0x20ccb60, 0xc0001e8840, 0x16, 0x22c19a0, 0xc0003317a0)
	/usr/local/go/src/text/template/exec.go:262 +0x142
text/template.(*state).walkIfOrWith(0xc000025868, 0xa, 0x20ccb60, 0xc0001e8840, 0x16, 0xc00059a840, 0xc0003316e0, 0xc0003317a0)
	/usr/local/go/src/text/template/exec.go:295 +0x222
text/template.(*state).walk(0xc000025868, 0x20ccb60, 0xc0001e8840, 0x16, 0x22c1960, 0xc0002bf6c0)
	/usr/local/go/src/text/template/exec.go:259 +0x327
text/template.(*state).walk(0xc000025868, 0x20ccb60, 0xc0001e8840, 0x16, 0x22c19a0, 0xc000331620)
	/usr/local/go/src/text/template/exec.go:262 +0x142
text/template.(*state).walkIfOrWith(0xc000025868, 0xa, 0x20ccb60, 0xc0001e8840, 0x16, 0xc00059a660, 0xc000331080, 0xc000331620)
	/usr/local/go/src/text/template/exec.go:295 +0x222
text/template.(*state).walk(0xc000025868, 0x20ccb60, 0xc0001e8840, 0x16, 0x22c1960, 0xc0002bf740)
	/usr/local/go/src/text/template/exec.go:259 +0x327
text/template.(*state).walk(0xc000025868, 0x20ccb60, 0xc0001e8840, 0x16, 0x22c19a0, 0xc0004ed080)
	/usr/local/go/src/text/template/exec.go:262 +0x142
text/template.(*Template).execute(0xc000481c40, 0x22a5000, 0xc0001323f0, 0x20ccb60, 0xc0001e8840, 0x0, 0x0)
	/usr/local/go/src/text/template/exec.go:217 +0x215
text/template.(*Template).Execute(0xc000481c40, 0x22a5000, 0xc0001323f0, 0x20ccb60, 0xc0001e8840, 0xc000290dd0, 0xc000290dd0)
	/usr/local/go/src/text/template/exec.go:200 +0x53
text/template.(*Template).ExecuteTemplate(0xc000025a88, 0x22a5000, 0xc0001323f0, 0x2110994, 0x14, 0x20ccb60, 0xc0001e8840, 0x0, 0x0)
	/usr/local/go/src/text/template/exec.go:186 +0x77
github.com/Azure/acs-engine/pkg/acsengine.(*TemplateGenerator).GenerateTemplate(0xc0002ae2b8, 0xc0001fd5c0, 0x21048c9, 0x9, 0x2298de0, 0x7, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/Azure/acs-engine/pkg/acsengine/template_generator.go:92 +0x43d
github.com/Azure/acs-engine/cmd.(*generateCmd).run(0xc000188300, 0xc000322900, 0xc000177f00)
	/go/src/github.com/Azure/acs-engine/cmd/generate.go:187 +0x212
github.com/Azure/acs-engine/cmd.newGenerateCmd.func1(0xc000322900, 0xc000177f00, 0x1, 0x1, 0x0, 0x0)
	/go/src/github.com/Azure/acs-engine/cmd/generate.go:60 +0x108
github.com/Azure/acs-engine/vendor/github.com/spf13/cobra.(*Command).execute(0xc000322900, 0xc000177ed0, 0x1, 0x1, 0xc000322900, 0xc000177ed0)
	/go/src/github.com/Azure/acs-engine/vendor/github.com/spf13/cobra/command.go:647 +0x3f1
github.com/Azure/acs-engine/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc000305d40, 0xc000323200, 0xc000322fc0, 0xc000322d80)
	/go/src/github.com/Azure/acs-engine/vendor/github.com/spf13/cobra/command.go:726 +0x2cc
github.com/Azure/acs-engine/vendor/github.com/spf13/cobra.(*Command).Execute(0xc000305d40, 0xc00000c018, 0x0)
	/go/src/github.com/Azure/acs-engine/vendor/github.com/spf13/cobra/command.go:685 +0x2b
main.main()
	/go/src/github.com/Azure/acs-engine/main.go:12 +0x74
```

**What you expected to happen**:
We should be able to turn off `ip-masq-agent` if not desired in the ACS cluster.

**How to reproduce it** (as minimally and precisely as possible):
Run `acs-engine generate <filename>.json` with the following block:

```
        "addons": [
          {
           "name": "ip-masq-agent",
           "enabled" : false
          }
        ]
```

**Anything else we need to know**:
NO

**Comment by jackfrancis** [comment]
This is probably a real issue.

1) ip-masq-agent is probably implemented in such a way that it is not really able to be disabled (i.e., it's a required addon)
2) there is probably no "if ip-masq-agent is disabled, apply the IPTABLES rules manually" implemented for vm nodes

#1 should definitely be rationalized. If we want to enable #2, then we should make the "disable" option implement #2. If we just want to require ip-masq-agent for all clusters as a strong opinion, then we should fix the validation code so that the shared failure doesn't occur in that way.

**Comment by tariq1890** [comment]
This should be fixed by #167 & #169
